File: backend/platform-service/internal/model/topics.go
package model

// TopicID представляет идентификатор темы для языка программирования.
type TopicID string

// Topic представляет тему для языка программирования.
type Topic struct {
	ID   TopicID
	Name string
}

// LanguageTopics содержит карту тем для каждого языка программирования.
var LanguageTopics = map[Language][]Topic{
	LanguagePython: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы данных"},
		{ID: TopicID("lists_arrays"), Name: "Списки и массивы"},
		{ID: TopicID("dictionaries"), Name: "Словари"},
		{ID: TopicID("functions"), Name: "Функции"},
		{ID: TopicID("closures"), Name: "Замыкания"},
		{ID: TopicID("decorators"), Name: "Декораторы"},
		{ID: TopicID("generators"), Name: "Генераторы"},
		{ID: TopicID("classes_oop"), Name: "Классы и ООП"},
		{ID: TopicID("exceptions"), Name: "Обработка исключений"},
		{ID: TopicID("context_managers"), Name: "Контекстные менеджеры"},
		{ID: TopicID("async_await"), Name: "Асинхронное программирование"},
	},
	LanguageJavaScript: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы"},
		{ID: TopicID("arrays"), Name: "Массивы"},
		{ID: TopicID("objects"), Name: "Объекты"},
		{ID: TopicID("functions"), Name: "Функции"},
		{ID: TopicID("closures"), Name: "Замыкания"},
		{ID: TopicID("this_binding"), Name: "Контекст выполнения (this)"},
		{ID: TopicID("prototypes"), Name: "Прототипы"},
		{ID: TopicID("classes"), Name: "Классы (ES6+)"},
		{ID: TopicID("promises_async"), Name: "Промисы и async/await"},
		{ID: TopicID("event_loop"), Name: "Event Loop"},
		{ID: TopicID("destructuring"), Name: "Деструктуризация"},
		{ID: TopicID("modules"), Name: "Модули (ES6+)"},
	},
	LanguageGo: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы"},
		{ID: TopicID("slices"), Name: "Срезы"},
		{ID: TopicID("maps"), Name: "Мапы"},
		{ID: TopicID("functions"), Name: "Функции"},
		{ID: TopicID("methods"), Name: "Методы"},
		{ID: TopicID("interfaces"), Name: "Интерфейсы"},
		{ID: TopicID("goroutines"), Name: "Горутины"},
		{ID: TopicID("channels"), Name: "Каналы"},
		{ID: TopicID("select"), Name: "Select statement"},
		{ID: TopicID("defer_panic_recover"), Name: "Defer, panic, recover"},
		{ID: TopicID("pointers"), Name: "Указатели"},
		{ID: TopicID("structs"), Name: "Структуры"},
	},
	LanguageJava: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы"},
		{ID: TopicID("arrays_lists"), Name: "Массивы и списки"},
		{ID: TopicID("collections"), Name: "Коллекции (Set, Map)"},
		{ID: TopicID("methods"), Name: "Методы"},
		{ID: TopicID("classes_objects"), Name: "Классы и объекты"},
		{ID: TopicID("inheritance"), Name: "Наследование"},
		{ID: TopicID("interfaces"), Name: "Интерфейсы"},
		{ID: TopicID("generics"), Name: "Дженерики"},
		{ID: TopicID("exceptions"), Name: "Исключения"},
		{ID: TopicID("streams"), Name: "Streams API"},
		{ID: TopicID("lambda_expressions"), Name: "Lambda выражения"},
		{ID: TopicID("concurrency"), Name: "Многопоточность"},
	},
	LanguageCpp: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы"},
		{ID: TopicID("pointers_references"), Name: "Указатели и ссылки"},
		{ID: TopicID("arrays_vectors"), Name: "Массивы и векторы"},
		{ID: TopicID("functions"), Name: "Функции"},
		{ID: TopicID("classes_objects"), Name: "Классы и объекты"},
		{ID: TopicID("inheritance"), Name: "Наследование"},
		{ID: TopicID("templates"), Name: "Шаблоны"},
		{ID: TopicID("smart_pointers"), Name: "Умные указатели"},
		{ID: TopicID("stl"), Name: "STL контейнеры и алгоритмы"},
		{ID: TopicID("move_semantics"), Name: "Move семантика"},
		{ID: TopicID("lambda"), Name: "Lambda выражения"},
		{ID: TopicID("multithreading"), Name: "Многопоточность"},
	},
	LanguageRust: {
		{ID: TopicID("variables_types"), Name: "Переменные и типы"},
		{ID: TopicID("ownership"), Name: "Владение (ownership)"},
		{ID: TopicID("borrowing"), Name: "Заимствование (borrowing)"},
		{ID: TopicID("lifetimes"), Name: "Время жизни"},
		{ID: TopicID("vectors"), Name: "Векторы"},
		{ID: TopicID("hashmaps"), Name: "HashMap"},
		{ID: TopicID("functions"), Name: "Функции"},
		{ID: TopicID("structs"), Name: "Структуры"},
		{ID: TopicID("enums"), Name: "Перечисления"},
		{ID: TopicID("pattern_matching"), Name: "Сопоставление с образцом"},
		{ID: TopicID("error_handling"), Name: "Обработка ошибок (Result, Option)"},
		{ID: TopicID("concurrency"), Name: "Многопоточность"},
	},
	LanguageTypeScript: {
		{ID: TopicID("types"), Name: "Типы"},
		{ID: TopicID("interfaces"), Name: "Интерфейсы"},
		{ID: TopicID("generics"), Name: "Дженерики"},
		{ID: TopicID("unions_intersections"), Name: "Объединения и пересечения типов"},
		{ID: TopicID("type_guards"), Name: "Защитники типов"},
		{ID: TopicID("decorators"), Name: "Декораторы"},
		{ID: TopicID("utility_types"), Name: "Утилитарные типы"},
		{ID: TopicID("modules"), Name: "Модули"},
		{ID: TopicID("async_promises"), Name: "Асинхронность"},
		{ID: TopicID("classes"), Name: "Классы"},
		{ID: TopicID("namespaces"), Name: "Пространства имен"},
	},
}

// String возвращает строковое представление идентификатора темы.
func (tid TopicID) String() string {
	return string(tid)
}

// GetTopicsForLanguage возвращает список тем для указанного языка.
func GetTopicsForLanguage(language Language) []Topic {
	topics, ok := LanguageTopics[language]
	if !ok {
		return nil
	}
	return topics
}

// IsValidTopic проверяет, является ли тема валидной для указанного языка.
func IsValidTopic(language Language, topic TopicID) bool {
	topics := GetTopicsForLanguage(language)
	if topics == nil {
		return false
	}

	for _, t := range topics {
		if t.ID == topic {
			return true
		}
	}

	return false
}

// IsValidTopicString проверяет, является ли тема валидной для указанного языка (для обратной совместимости).
func IsValidTopicString(language, topic string) bool {
	return IsValidTopic(Language(language), TopicID(topic))
}

File: backend/content-service/app/models/topics.py
"""Topic definitions for programming languages."""

from app.models.enums import Language
from pydantic import BaseModel


class Topic(BaseModel):
    """Topic model with ID and name."""

    topic_id: str
    name: str

    class Config:
        """Pydantic config."""

        populate_by_name = True


# Topic definitions for each language
LANGUAGE_TOPICS: dict[Language, list[Topic]] = {
    Language.PYTHON: [
        Topic(topic_id="variables_types", name="Переменные и типы данных"),
        Topic(topic_id="lists_arrays", name="Списки и массивы"),
        Topic(topic_id="dictionaries", name="Словари"),
        Topic(topic_id="functions", name="Функции"),
        Topic(topic_id="closures", name="Замыкания"),
        Topic(topic_id="decorators", name="Декораторы"),
        Topic(topic_id="generators", name="Генераторы"),
        Topic(topic_id="classes_oop", name="Классы и ООП"),
        Topic(topic_id="exceptions", name="Обработка исключений"),
        Topic(topic_id="context_managers", name="Контекстные менеджеры"),
        Topic(topic_id="async_await", name="Асинхронное программирование"),
    ],
    Language.JAVASCRIPT: [
        Topic(topic_id="variables_types", name="Переменные и типы"),
        Topic(topic_id="arrays", name="Массивы"),
        Topic(topic_id="objects", name="Объекты"),
        Topic(topic_id="functions", name="Функции"),
        Topic(topic_id="closures", name="Замыкания"),
        Topic(topic_id="this_binding", name="Контекст выполнения (this)"),
        Topic(topic_id="prototypes", name="Прототипы"),
        Topic(topic_id="classes", name="Классы (ES6+)"),
        Topic(topic_id="promises_async", name="Промисы и async/await"),
        Topic(topic_id="event_loop", name="Event Loop"),
        Topic(topic_id="destructuring", name="Деструктуризация"),
        Topic(topic_id="modules", name="Модули (ES6+)"),
    ],
    Language.GO: [
        Topic(topic_id="variables_types", name="Переменные и типы"),
        Topic(topic_id="slices", name="Срезы"),
        Topic(topic_id="maps", name="Мапы"),
        Topic(topic_id="functions", name="Функции"),
        Topic(topic_id="methods", name="Методы"),
        Topic(topic_id="interfaces", name="Интерфейсы"),
        Topic(topic_id="goroutines", name="Горутины"),
        Topic(topic_id="channels", name="Каналы"),
        Topic(topic_id="select", name="Select statement"),
        Topic(topic_id="defer_panic_recover", name="Defer, panic, recover"),
        Topic(topic_id="pointers", name="Указатели"),
        Topic(topic_id="structs", name="Структуры"),
    ],
    Language.JAVA: [
        Topic(topic_id="variables_types", name="Переменные и типы"),
        Topic(topic_id="arrays_lists", name="Массивы и списки"),
        Topic(topic_id="collections", name="Коллекции (Set, Map)"),
        Topic(topic_id="methods", name="Методы"),
        Topic(topic_id="classes_objects", name="Классы и объекты"),
        Topic(topic_id="inheritance", name="Наследование"),
        Topic(topic_id="interfaces", name="Интерфейсы"),
        Topic(topic_id="generics", name="Дженерики"),
        Topic(topic_id="exceptions", name="Исключения"),
        Topic(topic_id="streams", name="Streams API"),
        Topic(topic_id="lambda_expressions", name="Lambda выражения"),
        Topic(topic_id="concurrency", name="Многопоточность"),
    ],
    Language.CPP: [
        Topic(topic_id="variables_types", name="Переменные и типы"),
        Topic(topic_id="pointers_references", name="Указатели и ссылки"),
        Topic(topic_id="arrays_vectors", name="Массивы и векторы"),
        Topic(topic_id="functions", name="Функции"),
        Topic(topic_id="classes_objects", name="Классы и объекты"),
        Topic(topic_id="inheritance", name="Наследование"),
        Topic(topic_id="templates", name="Шаблоны"),
        Topic(topic_id="smart_pointers", name="Умные указатели"),
        Topic(topic_id="stl", name="STL контейнеры и алгоритмы"),
        Topic(topic_id="move_semantics", name="Move семантика"),
        Topic(topic_id="lambda", name="Lambda выражения"),
        Topic(topic_id="multithreading", name="Многопоточность"),
    ],
    Language.RUST: [
        Topic(topic_id="variables_types", name="Переменные и типы"),
        Topic(topic_id="ownership", name="Владение (ownership)"),
        Topic(topic_id="borrowing", name="Заимствование (borrowing)"),
        Topic(topic_id="lifetimes", name="Время жизни"),
        Topic(topic_id="vectors", name="Векторы"),
        Topic(topic_id="hashmaps", name="HashMap"),
        Topic(topic_id="functions", name="Функции"),
        Topic(topic_id="structs", name="Структуры"),
        Topic(topic_id="enums", name="Перечисления"),
        Topic(topic_id="pattern_matching", name="Сопоставление с образцом"),
        Topic(
            topic_id="error_handling",
            name="Обработка ошибок (Result, Option)",
        ),
        Topic(topic_id="concurrency", name="Многопоточность"),
    ],
    Language.TYPESCRIPT: [
        Topic(topic_id="types", name="Типы"),
        Topic(topic_id="interfaces", name="Интерфейсы"),
        Topic(topic_id="generics", name="Дженерики"),
        Topic(
            topic_id="unions_intersections",
            name="Объединения и пересечения типов",
        ),
        Topic(topic_id="type_guards", name="Защитники типов"),
        Topic(topic_id="decorators", name="Декораторы"),
        Topic(topic_id="utility_types", name="Утилитарные типы"),
        Topic(topic_id="modules", name="Модули"),
        Topic(topic_id="async_promises", name="Асинхронность"),
        Topic(topic_id="classes", name="Классы"),
        Topic(topic_id="namespaces", name="Пространства имен"),
    ],
}


def get_topics_for_language(language: Language) -> list[Topic]:
    """Get topics for a specific language."""
    return LANGUAGE_TOPICS.get(language, [])


def is_valid_topic(language: Language, topic: str) -> bool:
    """Check if topic is valid for the given language."""
    topics = LANGUAGE_TOPICS.get(language, [])
    return any(t.topic_id == topic for t in topics)
